### ISSUE TITLE
new user registration

### ISSUE THREAD

**Issue by MeghanHughes** [WhareHauora/App#52]
I'm just looking into the user flow for registering as a new user to get the setup process started.

Hey @Br3nda do we have a user table set up in the db? I can't see how to POST to it in the [postman docs](https://documenter.getpostman.com/view/3560032/wharehauora/RVnb9wYN#intro)

**Comment by MeghanHughes** [comment]
@Br3nda Another question: what is the purpose of the whānau page? or, rather, what is the purpose of signing up individual members of the whare?

**Comment by MeghanHughes** [comment]
Ok so here is a suggestion.

**Step 1:** 
![screenshot at 2018-04-26 13-51-18](https://user-images.githubusercontent.com/22287063/39281578-8533a600-4959-11e8-9ee4-414ab63399e9.png)

| User        | Client-side           | Back-end  |
| ------------- |:-------------:| -----:|
| gives email address and a password and accepts T&C     | renders ![screenshot at 2018-04-26 13-58-20](https://user-images.githubusercontent.com/22287063/39281650-e8259642-4959-11e8-9ffe-31fafdd0462f.png) | creates a client_id, userName and password |

**Comment by MeghanHughes** [comment]
**Step 2:** 
| User        | Client-side           | Back-end  |
| ------------- |:-------------:| -----:|
| goes to email and clicks on confirmation link     | redirects to a setup wizard | -- |

We could greet them on the page with a stepper summary of what they need to set up, like
![screenshot at 2018-04-26 13-33-07](https://user-images.githubusercontent.com/22287063/39281920-377cd8b2-495b-11e8-9b24-eacc2bad8ab9.png)
With: step 1 - Account profile; step 2 - Add whare; step 3 - Add rooms and pair sensors.

**Comment by MeghanHughes** [comment]
**Step 3:** 

| User        | Client-side           | Back-end  |
| ------------- |:-------------:| -----:|
| finishes Account profile info | renders ![screenshot at 2018-04-26 13-54-02](https://user-images.githubusercontent.com/22287063/39282077-fb94dcb8-495b-11e8-884e-9def48f85447.png) | adds first_name, last_name |

**Comment by MeghanHughes** [comment]
**Step 4:** 

| User        | Client-side           | Back-end  |
| ------------- |:-------------:| -----:|
| adds Whare info | renders ![screenshot at 2018-04-26 13-18-49](https://user-images.githubusercontent.com/22287063/39282145-3b12244a-495c-11e8-8346-6a189e9634ae.png)| adds whare_name, property_type |

**Comment by MeghanHughes** [comment]
**Step 5:** 

| User        | Client-side           | Back-end  |
| ------------- |:-------------:| -----:|
| adds room info | renders ![screenshot at 2018-04-26 13-52-58](https://user-images.githubusercontent.com/22287063/39282186-7afaae24-495c-11e8-8c1e-0f5e0509b772.png) | adds room_name, room_type |

**Comment by MeghanHughes** [comment]
**Step 7:** 
Maybe have a spinner while data is being collected and rendered.

| User        | Client-side           | Back-end  |
| ------------- |:-------------:| -----:|
| can review info in dashboard | renders
![screenshot at 2018-04-26 14-25-25](https://user-images.githubusercontent.com/22287063/39282500-bac3ac30-495d-11e8-98c3-0543dc7f42b1.png)
 | adds first data set into the sensor history |

**Comment by MeghanHughes** [comment]
User will also be able to navigate to room profile pages BUT there will be limited information available until a history of sensor readings is created. 
![screenshot at 2018-04-26 14-28-49](https://user-images.githubusercontent.com/22287063/39282616-31129540-495e-11e8-9783-38c7d8d4e150.png)

**Comment by Br3nda** [comment]
The purpose is for each user to have their own account, rather than sharing a password.
Think of it as someone sharing the account with their partner, but later that partner leaves the relationship. The primary account holder than add/remove that partner's account from the home if she no longer wants them to know the environmental conditions in their home. Because you can extrapolate whether someone is home or not from their use of the heating etc, it's important people have control over who can see it.

**Comment by MeghanHughes** [comment]
That's a really good point!!
So - one whare can have many members and each member get access to the same account information (of that whare). And they all get the same admin access to changing details about the whare profile (which includes rooms and whanau members) And so they can also all decide on what notifications they want - some might not want any (if, maybe, there is just one person who is responsible for actioning on all the alerts so they are the one to get alert notifications).

**Comment by MeghanHughes** [comment]
@Br3nda do you think a good flow would be to set up a one-off setup process, so that the personal account profile, whare, room and whanau components are all on a single page (rather than account on one page, whare+rooms on one page, whanau on a separate page)? Alternatively, one can flow on to another in the setup process, with a stepper to show progress.

**Comment by Br3nda** [comment]
All in one page sounds good to me. I had only separated them because i buokt the data model that way, which is not a good reason.

**Comment by MeghanHughes** [comment]
Ok I am going to start breaking this down a little
- [ ] set up CRUD routes @Br3nda (or are they already set up?)
- [ ] design front end page for initial setup form (Meghan)
- [ ] create email confirmation and link to set up form (@Br3nda how does the server handle this? Ruby has some kind of mailer gem?)
- [ ] connect the datapoints to db (Meghan and @anna-marie-vos - once we have details about the user table and POST route)
- [ ] how does the sensor pairing happen? @Br3nda 
- [ ] decide on validation requirements: new user must set up: at least one whare; at least one room; at least one paired sensor in order to send form (I can set up validation methods after I have made the form components)
- [ ] storing cookies/tokens and refreshing the tokens - I'm not sure how this happens (@Br3nda ?)

**Comment by Br3nda** [comment]
Yes, the server will email on user creation.

Currently, you only need to pair the gateway to your account, not each and every sensor.
You tell a sensor what account it is to use when you turn it on.

There's the login example in the PR i sent. Once you have a session, you get both a cookie and a token in the response. Either is good enough to use on the next request.

**Comment by MeghanHughes** [comment]
Ok so I have set up a one page initial set up page on [this branch](https://github.com/WhareHauora/App/tree/mh/initialSetUp)
and you can view it on route `/set-up`
I have added [Final Form](https://github.com/final-form/) and use that to build the react form components.
At the moment, the user inputs are held in state in state>values. I have set up each component with it's own react form (I don't know if this is the best way to do this - I will get Iain to review it) but it still means that ALL user inputs persist in state so I am thinking the submission logic just needs to grab those bits, validate them, and then pass them at once over to the db. 
My reasoning behind setting it up like this is I am pretty sure each component is sending data to a different data table so actually the submission logic is sending several calls to the api to POST the new info. 
It has thrown up a new issue in the `/account` route though - which I will make an issue for.

**Comment by Br3nda** [comment]
I don't think there are API methods for new registrations yet. We were thinking we give people an account when we give them sensors (there's was no need to have every person who found us on the internet sign up when they don't have a device).

@Kamikazilady is that still the plan/thinking. Do you want full user sign up?

**Comment by MeghanHughes** [comment]
Ah ok - the page is (pretty much - needs more styling) set up so in order to get a full user sign up working we would just need the CRUD routes and connect the inputs to the db tables. 
But also - are their POST, UPDATE and DELETE routes set up for the rooms, whare and whanau?

**Comment by Kamikazilady** [comment]
The user should be able to sign up when they receive their sensors.
Everything about the process as per previous emails is good.

I'll run a "newsletter" sign up separately, does that answer the question?

I did realise though that we're not capturing housing building types ie
brick, cladding etc but at this stage we don't know it and it's gonna need
a lot of analysis and ux to be easy for the user to choose. So I'll leave
out.

**Comment by MeghanHughes** [comment]
@Kamikazilady what is the newsletter sign up?

**Comment by MeghanHughes** [comment]
Hi @Br3nda and @Kamikazilady can I just clarify the need for create, update and delete methods for the rooms, whare and whanau pages. 
We can read these out of the database, but if we are to let users add, edit and delete these we need to set up CUD routes to the server?

**Comment by Kamikazilady** [comment]
Heya,

Yes we're going to need to, otherwise it's a one shot deal.

Can you estimate the time needed and we'll talk about this on Monday?

Then we can make plan from there on the work we still need to get done,
cool?

I don't know how to tell Janie and Anna Marie that. Could you

My apologies, we needed to set the scope better, it's obvious we've gone
over, i should have kept a closer eye, its cool and brilliant work from
Manu but we should bring it in and set out boundaries.

Ka pai?
Na
Hīria

**Comment by MeghanHughes** [comment]
All good @Kamikazilady !

We are so very close! We got held up on an bug that popped up on Friday but we should have that fixed by tomorrow. 

Ka kite anō!

**Comment by MeghanHughes** [comment]
Also this one @iainkirkpatrick - longer to read but further discussion on a user table
